### PR TITLE
Configure flake8 Linter and pre-commit 

### LINKED ISSUE #13: As a developer I am not let to commit with PEP-8 being followed
It's very important to follow coding standards, specially those which are made by the community. The problem statement here is to make sure everyone follows the basic standards, and as humans, we tend to forget.

Hence, we can automate this by applying linters as a pre-commit hook.

The task here is to add in README the following:

- installation instruction for Flake8
- procedure to set a pre-commit hook for Flake8

### PR DESCRIPTION
**Pull request by vinaypuppal**
# What kind of change does this PR introduce
closes #13 

## What is the current behavior
No linter in project

## What is the new behavior
- Configured [Flake8](https://medium.com/python-pandemonium/what-is-flake8-and-why-we-should-use-it-b89bd78073f2) linter
- Added and configured [pre-commit](http://pre-commit.com/) so `flake8` is run for every `git commit`
- Updated `Contributing.md` on how to install `pre-commit` hooks
- Example screenshot of `pre-commit` hook 
screen shot 2017-12-01 at 02 03 36

### Checklist for this PR

* [x] I have read [contributing guidelines](https://github.com/coderplex/coderplex-backend/blob/develop/.github/CONTRIBUTING.md).
* [x] Updated the requirements (`pip freeze > requirements/requirements.txt`). 
* [N/A] Created new migrations (`python webapp/manage.py makemigrations && python webapp/manage.py migrate`).
* [x] Pull request is towards `develop` branch.
* [x] Documentation
* [x] Ready to be merged. 

### Whats missing

- I did not actually fix all linting errors because I am not familiar with whole code base and they are many 😃. When I ran `flake8` I got many lint errors which can found below.
- Travis CI configuration for lint. @pbteja1998 Use [this info](http://pre-commit.com/#usage-in-continuous-integration) for setting up `pre-commit` in travis config

<details>
<summary>Lint Errors</summary>

```
Flake8..............................................................
.....Failedhookid: flake8

webapp/books/tests.py:1:1: F401 'django.test.TestCase' imported but
unused
webapp/authentication/views.py:2:80: E501 line too long (87 > 79 cha
racters)
webapp/authentication/views.py:6:1: F401 'rest_framework.permissions
' imported but unused
webapp/authentication/views.py:8:1: E302 expected 2 blank lines, fou
nd 1webapp/authentication/views.py:13:1: E302 expected 2 blank lines, fo
und 1
webapp/authentication/views.py:16:50: W292 no newline at end of file
webapp/books/urls.py:1:1: F401 'django.conf.urls.include' imported b
ut unused
webapp/books/urls.py:5:1: W293 blank line contains whitespace
webapp/books/urls.py:7:1: W293 blank line contains whitespace
webapp/books/urls.py:9:80: E501 line too long (91 > 79 characters)
webapp/books/urls.py:10:80: E501 line too long (91 > 79 characters)
webapp/books/urls.py:11:1: W293 blank line contains whitespacewebapp/books/urls.py:12:1: W293 blank line contains whitespace
webapp/books/urls.py:15:80: E501 line too long (88 > 79 characters)
webapp/books/urls.py:18:2: W292 no newline at end of file
webapp/books/migrations/0001_initial.py:22:80: E501 line too long (1
14 > 79 characters)
webapp/books/migrations/0001_initial.py:25:80: E501 line too long (9
6 > 79 characters)
webapp/books/migrations/0001_initial.py:29:80: E501 line too long (1
24 > 79 characters)
webapp/books/migrations/0001_initial.py:30:80: E501 line too long (1
54 > 79 characters)
webapp/books/migrations/0001_initial.py:36:80: E501 line too long (1
14 > 79 characters)
webapp/books/migrations/0001_initial.py:40:80: E501 line too long (8
3 > 79 characters)
webapp/books/migrations/0001_initial.py:43:80: E501 line too long (1
31 > 79 characters)
webapp/books/migrations/0001_initial.py:44:80: E501 line too long (1
24 > 79 characters)
webapp/books/migrations/0001_initial.py:45:80: E501 line too long (1
56 > 79 characters)
webapp/books/migrations/0001_initial.py:51:80: E501 line too long (1
14 > 79 characters)
webapp/books/migrations/0001_initial.py:55:80: E501 line too long (8
3 > 79 characters)
webapp/books/migrations/0001_initial.py:58:80: E501 line too long (1
37 > 79 characters)webapp/books/migrations/0001_initial.py:59:80: E501 line too long (1
24 > 79 characters)
webapp/books/migrations/0001_initial.py:60:80: E501 line too long (1
54 > 79 characters)
webapp/coderplex_apis/config/localhost.py:19:63: E231 missing whites
pace after ','
webapp/coderplex_apis/config/localhost.py:19:80: E501 line too long
(96 > 79 characters)
webapp/coderplex_apis/config/localhost.py:20:80: E501 line too long
(93 > 79 characters)
webapp/coderplex_apis/config/localhost.py:20:94: W292 no newline atend of file
webapp/member/models.py:5:1: F401 'django.contrib.auth.login' import
ed but unused
webapp/member/models.py:7:1: E302 expected 2 blank lines, found 1
webapp/member/models.py:14:80: E501 line too long (88 > 79 character
s)
webapp/member/models.py:15:80: E501 line too long (85 > 79 characters)
webapp/member/models.py:17:80: E501 line too long (87 > 79 characters)
webapp/member/models.py:18:80: E501 line too long (82 > 79 characters)
webapp/member/models.py:19:80: E501 line too long (88 > 79 characters)
webapp/member/models.py:20:80: E501 line too long (90 > 79 characters)
webapp/member/models.py:21:80: E501 line too long (89 > 79 characters)
webapp/member/models.py:22:80: E501 line too long (90 > 79 characters)
webapp/member/models.py:23:80: E501 line too long (89 > 79 characters)
webapp/member/models.py:24:80: E501 line too long (90 > 79 characters)
webapp/member/models.py:25:80: E501 line too long (97 > 79 characters)
webapp/member/models.py:26:80: E501 line too long (99 > 79 characters)
webapp/member/models.py:28:1: E302 expected 2 blank lines, found 1
webapp/member/models.py:33:25: E231 missing whitespace after ','
webapp/member/models.py:36:9: F841 local variable 'avatar_url' is assigned to but never used
webapp/member/models.py:37:21: E221 multiple spaces before operator
webapp/member/models.py:56:63: W292 no newline at end of file
webapp/authentication/models.py:1:1: F401 'django.db.models' imported but unused
webapp/authentication/tests.py:1:1: F401 'django.test.TestCase' imported but unused
webapp/member/tests.py:1:1: F401 'django.test.TestCase' imported but unused
webapp/manage.py:14:13: F401 'django' imported but unused
webapp/books/serializers.py:9:1: W293 blank line contains whitespace
webapp/books/serializers.py:14:1: W293 blank line contains whitespace
webapp/books/serializers.py:19:1: W293 blank line contains whitespace
webapp/books/serializers.py:29:1: W293 blank line contains whitespace
webapp/books/serializers.py:39:1: W293 blank line contains whitespace
webapp/books/serializers.py:56:1: W293 blank line contains whitespace
webapp/books/serializers.py:71:1: W293 blank line contains whitespace
webapp/books/serializers.py:72:1: W293 blank line contains whitespace
webapp/books/serializers.py:77:1: W293 blank line contains whitespace
webapp/coderplex_apis/config/production.py:7:18: E201 whitespace after '['
webapp/coderplex_apis/config/production.py:10:1: E265 block commentshould start with '# '
webapp/coderplex_apis/config/production.py:24:60: W292 no newline at end of file
webapp/authentication/serializers.py:3:1: E302 expected 2 blank lines, found 1
webapp/authentication/serializers.py:6:1: E302 expected 2 blank lines, found 1
webapp/authentication/serializers.py:7:49: W292 no newline at end of file
webapp/member/admin.py:4:1: E302 expected 2 blank lines, found 1
webapp/member/admin.py:8:1: E305 expected 2 blank lines after classor function definition, found 1
webapp/member/admin.py:8:54: W292 no newline at end of file
webapp/member/views.py:2:80: E501 line too long (115 > 79 characters)
webapp/member/views.py:8:1: E302 expected 2 blank lines, found 1
webapp/member/views.py:26:1: E302 expected 2 blank lines, found 1
webapp/member/views.py:27:22: E128 continuation line under-indentedfor visual indent
webapp/member/views.py:33:80: E501 line too long (84 > 79 characters)
webapp/member/views.py:38:80: E501 line too long (84 > 79 characters)
webapp/member/views.py:45:47: W292 no newline at end of file
webapp/books/models.py:11:1: W293 blank line contains whitespace
webapp/books/models.py:16:1: W293 blank line contains whitespace
webapp/books/models.py:19:1: W293 blank line contains whitespace
webapp/books/models.py:20:1: W293 blank line contains whitespace
webapp/books/models.py:43:1: W293 blank line contains whitespace
webapp/books/models.py:48:1: W293 blank line contains whitespace
webapp/books/models.py:50:26: W292 no newline at end of file
webapp/books/views.py:1:1: F401 'django.shortcuts.render' imported but unused
webapp/books/views.py:7:1: F401 'rest_framework.views.APIView' imported but unused
webapp/books/views.py:10:1: F401 'django.db.models.query.QuerySet' imported but unused
webapp/books/views.py:24:1: W293 blank line contains whitespace
webapp/books/views.py:27:21: E128 continuation line under-indented for visual indent
webapp/books/views.py:31:1: W293 blank line contains whitespace
webapp/books/views.py:37:21: E128 continuation line under-indented for visual indent
webapp/books/views.py:40:1: W293 blank line contains whitespace
webapp/books/views.py:45:1: W293 blank line contains whitespace
webapp/books/views.py:51:25: E127 continuation line over-indented for visual indent
webapp/books/views.py:54:1: W293 blank line contains whitespace
webapp/books/views.py:59:1: W293 blank line contains whitespace
webapp/books/views.py:62:1: W293 blank line contains whitespace
webapp/books/views.py:63:1: W293 blank line contains whitespace
webapp/books/views.py:69:1: W391 blank line at end of file
webapp/books/admin.py:14:54: W292 no newline at end of file
webapp/coderplex_apis/views.py:4:1: E302 expected 2 blank lines, found 1
webapp/coderplex_apis/views.py:10:1: E302 expected 2 blank lines, found 1
webapp/coderplex_apis/views.py:13:68: W292 no newline at end of file
webapp/coderplex_apis/settings.py:29:5: F403 'from .config.production import *' used; unable to detect undefined names
webapp/coderplex_apis/settings.py:29:5: F401 '.config.production.*'imported but unused
webapp/coderplex_apis/settings.py:31:5: F403 'from .config.localhost import *' used; unable to detect undefined names
webapp/coderplex_apis/settings.py:31:5: F401 '.config.localhost.*' imported but unused
webapp/coderplex_apis/settings.py:44:1: W293 blank line contains whitespace
webapp/coderplex_apis/settings.py:100:80: E501 line too long (91 > 79 characters)
webapp/coderplex_apis/settings.py:103:80: E501 line too long (81 > 79 characters)
webapp/coderplex_apis/settings.py:106:80: E501 line too long (82 > 79 characters)
webapp/coderplex_apis/settings.py:109:80: E501 line too long (83 > 79 characters)
webapp/coderplex_apis/settings.py:190:30: E203 whitespace before ':'
webapp/coderplex_apis/settings.py:201:60: W292 no newline at end offile
webapp/authentication/admin.py:1:1: F401 'django.contrib.admin' imported but unused
webapp/member/serializers.py:4:1: E302 expected 2 blank lines, found 1
webapp/member/serializers.py:22:80: E501 line too long (281 > 79 characters)
webapp/member/serializers.py:24:1: E302 expected 2 blank lines, found 1
webapp/member/serializers.py:31:80: E501 line too long (83 > 79 characters)
webapp/member/serializers.py:33:80: E501 line too long (80 > 79 characters)
webapp/member/serializers.py:38:1: E302 expected 2 blank lines, found 1
webapp/member/serializers.py:42:27: E231 missing whitespace after ','
webapp/member/serializers.py:42:43: E231 missing whitespace after ','
webapp/member/serializers.py:42:55: E231 missing whitespace after ','
webapp/member/serializers.py:42:68: E231 missing whitespace after ','
webapp/member/serializers.py:42:80: E501 line too long (260 > 79 characters)
webapp/member/serializers.py:42:83: E231 missing whitespace after ','
webapp/member/serializers.py:42:101: E231 missing whitespace after ','
webapp/member/serializers.py:42:118: E231 missing whitespace after ','
webapp/member/serializers.py:42:137: E231 missing whitespace after ','
webapp/member/serializers.py:42:155: E231 missing whitespace after ','
webapp/member/serializers.py:42:174: E231 missing whitespace after ','
webapp/member/serializers.py:42:192: E231 missing whitespace after ','
webapp/member/serializers.py:42:210: E231 missing whitespace after ','
webapp/member/serializers.py:42:234: E231 missing whitespace after ','
webapp/member/serializers.py:46:80: E501 line too long (92 > 79 characters)
webapp/member/serializers.py:47:80: E501 line too long (80 > 79 characters)
webapp/member/serializers.py:48:80: E501 line too long (83 > 79 characters)
webapp/member/serializers.py:49:80: E501 line too long (89 > 79 characters)
webapp/member/serializers.py:50:80: E501 line too long (98 > 79 characters)
webapp/member/serializers.py:51:80: E501 line too long (95 > 79 characters)
webapp/member/serializers.py:52:80: E501 line too long (101 > 79 characters)
webapp/member/serializers.py:53:80: E501 line too long (98 > 79 characters)
webapp/member/serializers.py:54:80: E501 line too long (101 > 79 characters)
webapp/member/serializers.py:55:80: E501 line too long (98 > 79 characters)
webapp/member/serializers.py:56:80: E501 line too long (98 > 79 characters)
webapp/member/serializers.py:57:80: E501 line too long (116 > 79 characters)
webapp/member/serializers.py:58:80: E501 line too long (122 > 79 characters)
webapp/member/serializers.py:60:24: W292 no newline at end of file
webapp/member/migrations/0001_initial.py:22:80: E501 line too long (114 > 79 characters)
webapp/member/migrations/0001_initial.py:23:80: E501 line too long (83 > 79 characters)
webapp/member/migrations/0001_initial.py:24:80: E501 line too long (104 > 79 characters)
webapp/member/migrations/0001_initial.py:25:80: E501 line too long (101 > 79 characters)
webapp/member/migrations/0001_initial.py:27:80: E501 line too long (103 > 79 characters)
webapp/member/migrations/0001_initial.py:29:80: E501 line too long (104 > 79 characters)
webapp/member/migrations/0001_initial.py:30:80: E501 line too long (106 > 79 characters)
webapp/member/migrations/0001_initial.py:31:80: E501 line too long (105 > 79 characters)
webapp/member/migrations/0001_initial.py:32:80: E501 line too long (106 > 79 characters)
webapp/member/migrations/0001_initial.py:33:80: E501 line too long (105 > 79 characters)
webapp/member/migrations/0001_initial.py:34:80: E501 line too long (106 > 79 characters)
webapp/member/migrations/0001_initial.py:35:80: E501 line too long (113 > 79 characters)
webapp/member/migrations/0001_initial.py:36:80: E501 line too long (115 > 79 characters)
webapp/member/migrations/0001_initial.py:37:80: E501 line too long (142 > 79 characters)
webapp/member/urls.py:1:1: F401 'django.conf.urls.include' importedbut unused
webapp/member/urls.py:7:2: W292 no newline at end of file
```

</details>

### CODE REVIEW

**Comment by vinaypuppal** [comment]
And also I updated `.gitignore` with more generalised ignores for all editors and python recommended ignores

**Comment by vinaypuppal** [comment]
Just now found [yapf](https://github.com/google/yapf) by google a python code formatter. I configured this with `pre-commit` and lint errors reduced 🎉. 

Below are remaining errors

```
hookid: flake8
webapp/coderplex_apis/config/production.py:10:1: E265 block comment
should start with '# '
webapp/authentication/views.py:2:80: E501 line too long (87 > 79 characters)
webapp/authentication/views.py:6:1: F401 'rest_framework.permissions
' imported but unused
webapp/member/models.py:5:1: F401 'django.contrib.auth.login' import
ed but unused
webapp/member/models.py:50:9: F841 local variable 'avatar_url' is as
signed to but never used
webapp/books/urls.py:1:1: F401 'django.conf.urls.include' imported b
ut unused
webapp/member/views.py:2:80: E501 line too long (115 > 79 characters
)
webapp/manage.py:14:13: F401 'django' imported but unused
webapp/coderplex_apis/settings.py:27:5: F403 'from .config.productio
n import *' used; unable to detect undefined names
webapp/coderplex_apis/settings.py:27:5: F401 '.config.production.*'
imported but unused
webapp/coderplex_apis/settings.py:29:5: F403 'from .config.localhost
 import *' used; unable to detect undefined names
webapp/coderplex_apis/settings.py:29:5: F401 '.config.localhost.*' imported but unused
webapp/coderplex_apis/settings.py:82:80: E501 line too long (83 > 79 characters)
webapp/books/views.py:1:1: F401 'django.shortcuts.render' imported but unused
webapp/books/views.py:7:1: F401 'rest_framework.views.APIView' imported but unused
webapp/books/views.py:10:1: F401 'django.db.models.query.QuerySet' imported but unused
```

**Comment by vinaypuppal** [comment]
Code Indentation/Formatting changed in PR is default rules set by yapf and above errors are also from default config of flake8.
Both flake8 and yapf can be configured with custom style preferences so check below links and configure as you prefer @pbteja1998 or @rrmerugu 
- http://flake8.pycqa.org/en/latest/
- https://github.com/google/yapf

File: .github/CONTRIBUTING.md [hunk 1]
```diff
@@ -95,11 +95,15 @@ We welcome pull requests from beginners and django developers alike!
...
    ```bash
     source venv/bin/activate
-   ```   
+   ```
 1. Install the requirements
    ```bash
     pip install -r requirements/requirements.txt
    ```
+1. Install pre-commit into your git hooks. pre-commit will now run on every commit. Every time you clone a project that is using pre-comm it running `pre-commit install` should always be the first thing you do after installing requirements
+   ```bash
```
Comment (pbteja1998) [inline .github/CONTRIBUTING.md]: There is a very small typo in the above line `pre-comm it running`  --> `pre-commit running`

Comment (pbteja1998) [inline .github/CONTRIBUTING.md]: There is a very small typo in the above line `pre-comm it running` --> `pre-commit running`

**Review by pbteja1998 — CHANGES_REQUESTED** [review]

**Review by pbteja1998 — DISMISSED** [review]

**Comment by pbteja1998** [comment]
I think this cannot be merged until #21 is merged as it should pull `.travis.yml` for passing the `travis-ci` checks

**Comment by vinaypuppal** [comment]
@pbteja1998 No need we can ignore status check and merge if this PR is ready to be merged. Actually we need to merge this PR first and fix above lint errors and then configure travis with pre-commit hooks otherwise build will fail due to lint errors.

**Review by pbteja1998 — APPROVED** [review]

**Comment by pbteja1998** [comment]
@vinaypuppal How can we ignore status checks, `Merge pull request` button is greyed out as travis-ci has failed.

**Comment by vinaypuppal** [comment]
@pbteja1998 I just disabled status checks requirements on PR's merging. We can reenable them once we configure travis fully.

**Pull request merged**